### PR TITLE
Support Google Gemini through AI Studio

### PR DESCRIPTION
**Pull request by rafax**
This PR adds support for Google Gemini models as exposed by AI Studio API.
Slack [context](https://sourcegraph.slack.com/archives/C072XJ7TT5G/p1716393829253889).

Closes AI-50.

## Test plan

- tested locally
- e2e tests

### CODE REVIEW

File: cmd/cody-gateway/shared/config/config.go [hunk 1]
```diff
@@ -309,6 +317,16 @@ func (c *Config) Load() {
...
 
 	c.Environment = c.Get("CODY_GATEWAY_ENVIRONMENT", "dev", "Environment name.")
+
+	c.Google.AccessToken = c.GetOptional("CODY_GATEWAY_GOOGLE_ACCESS_TOKEN", "The Google AI Studio access token to be used.")
+	c.Google.AllowedModels = splitMaybe(c.Get("CODY_GATEWAY_GOOGLE_ALLOWED_MODELS",
+		strings.Join([]string{
+			"gemini-1.5-pro-latest",
+			"gemini-1.5-flash-latest",
+		}, ","),
+		"OpenAI models that can to be used."),
```
Comment (janhartman) [inline cmd/cody-gateway/shared/config/config.go:327]:
```suggestion
		"Google models that can be used."),
```

**Review by janhartman — APPROVED** [review]

File: cmd/cody-gateway/internal/httpapi/completions/upstream.go [hunk 2]
```diff
@@ -64,7 +64,7 @@ var hopHeaders = map[string]struct{}{
 // Methods do not need to be concurrency-safe, as they are only called sequentially.
 type upstreamHandlerMethods[ReqT UpstreamRequest] interface {
 	// getAPIURLByFeature returns the upstream API endpoint to call for the given feature.
-	getAPIURLByFeature(codygateway.Feature) string
+	getAPIURL(codygateway.Feature, ReqT) string
```
Comment (vdavid) [inline cmd/cody-gateway/internal/httpapi/completions/upstream.go:67]: I see that the implementations of this method always ignore the second parameter, except the new `GoogleHandler`, but that only uses `req.ShouldStream()` (a bool) and req.Model (a string).
It feels meh. But the best alternative I could come up with it to use the `feature` argument and concat the two values as a string with some separator, but that's even uglier. So 👍 , just wanted to note that I thought about it and ack that this is the best solution we found without major changes.

**Review by vdavid — APPROVED** [review]
Looks 👌

File: cmd/cody-gateway/internal/httpapi/completions/upstream.go (hunk 2 above)
Comment (rafax) [reply cmd/cody-gateway/internal/httpapi/completions/upstream.go:67]: UpstreamRequest only has 3 properties (`GetModel()` `ShouldStream()` and `BuildPrompt()`, so I decided that Google using 2/3 is close enough to pass the whole request.

File: cmd/cody-gateway/shared/config/config.go (hunk 1 above)
Comment (rafax) [reply cmd/cody-gateway/shared/config/config.go]: Applied manually

**Pull request merged**